pkg/allocateip/allocateip.go

```diff
@@ -93,25 +99,54 @@ func ensureHostTunnelAddress(ctx context.Context, c client.Interface, nodename s
 		logCtx.WithError(err).Fatalf("Unable to retrieve tunnel address. Error getting node '%s'", nodename)
 	}
 
-	// Get the address
+	// Get the address and ipam attribute string
 	var addr string
+	var attrString string
 	if vxlan {
 		addr = node.Spec.IPv4VXLANTunnelAddr
+		attrString = vxlanIPAMAttrString
 	} else if node.Spec.BGP != nil {
 		addr = node.Spec.BGP.IPv4IPIPTunnelAddr
+		attrString = ipipIPAMAttrString
 	}
 
+	// Work out if we need to assign a tunnel address.
+	var assign, release bool
 	if addr == "" {
 		// The tunnel has no IP address assigned, assign one.
-		logCtx.Debug("tunnel is not assigned - assign IP")
-		assignHostTunnelAddr(ctx, c, nodename, cidrs, vxlan)
-	} else if isIpInPool(addr, cidrs) {
-		// The tunnel address is still valid, so leave as it.
-		logCtx.WithField("IP", addr).Info("tunnel address is still valid")
-	} else {
+		logCtx.WithField("Node", nodename).Info("Assign a new tunnel address")
+		release = false
+		assign = true
+	} else if !isIpInPool(addr, cidrs) {
 		// The address that is currently assigned is no longer part
 		// of an encapsulatin-enabled pool, so release the IP, and reassign.
-		logCtx.WithField("IP", addr).Info("Reassigning tunnel address")
+		logCtx.WithField("currentAddr", addr).Info("Current address is not in ippool, reassign")
+		release = true
+		assign = true
+	} else {
+		// The tunnel address is still part of an an encapsulatin-enabled pool.
+		ipAddr := gnet.ParseIP(addr)
+		if err != nil {
+			logCtx.WithError(err).Fatalf("Failed to parse the CIDR '%s'", addr)
+		}
+
+		// Check if we got correct assignment attributes
+		attr, attrErr := c.IPAM().GetAssignmentAttributes(ctx, net.IP{IP: ipAddr})
+		if attrErr == nil && attr[ipam.AttributeType] == attrString {
```

Comment:
nit: I'd just call this `err` - I don't think there is a need to call it something else unless we're dealing with two errors at the same time?

Reply:
Also, I think we need to explicitly handle this error. If this is an "Unable to reach data store error" we shouldn't fall into the `else` block and assign a new address. We only want to do that if it's an "Object doesn't exist" error.

Reply:
Agreed. However, it seems current ipam code lack of the ability to return correct error status. 
- https://github.com/projectcalico/libcalico-go/blob/master/lib/ipam/ipam.go#L1370
This line stopped caller from getting actual error code from datastore.  Maybe we should just return the error we got from datastore? 

- https://github.com/projectcalico/libcalico-go/blob/master/lib/ipam/ipam_block.go#L351
This line is the same. 

`GetAssignmentAtrributes` seems not been used by anyone? Should I modify this function to return correct status?

Reply:
Yeah, the first one should just return the error from the data store - I agree.

The second should probably be an `ErrorResourceDoesNotExist`.

Then, in this code we can just check for `ErrorResourceDoesNotExist` here. WDYT?

Reply:
Sounds good. I will raise a PR tomorrow on libcalico-go repo.

Reply:
libcalico-go PR has been merged. I have got UT cover all cases. 
I implemented a `shimClient` to generate random ipam datastore errors. The code is a bit excessive but I cannot find any other way to do it easily.